KOGITO-7303 - Added "Basic Integration Test with REST Assured" guide (#11)
kiegroup/kogito-docs: serverlessworkflow/modules/ROOT/pages/testing-and-troubleshooting/basic-integration-tests-with-restassured.adoc

> = Basic Integration Test with REST Assured

Reviewer: maybe add a header like this https://asciidoctor.org/docs/asciidoc-recommended-practices/#document-title ?

Author: Sorry @spolti. You mean the syntax, the author name, or something else?

Reviewer: the author name is not needed, just the metadata and the link, similar to https://github.com/kiegroup/kogito-docs/pull/8/files#diff-db2569701a9160b1b430536d457e6df1f77ef0ad4a9f26edd8ad440ea9ed9558R1

As a developer, you can use https://rest-assured.io/[REST Assured] to test your Serverless Workflow application,
which includes the following procedures:

* <<boostrapping-the-project,Bootstrapping a project>>
* <<creating-workflow,Creating a workflow>>
* <<testing-application,Testing your Serverless Workflow application>>

Also, you can directly access an example application by cloning `kogito-quarkus-examples/serverless-workflow-testing-with-rest-assured`

> from the https://github.com/kiegroup/kogito-examples/tree/main/kogito-quarkus-examples/serverless-workflow-testing-with-rest-assured[GitHub repository].

Reviewer 2: I think the link is correct. 🤔

+
--
.Create a project using Quarkus CLI
[source,shell]
----
quarkus create app \
    -x=kogito-quarkus-serverless-workflow \
    -x=quarkus-resteasy-jackson \
    org.kie.kogito.examples:serverless-workflow-testing-with-rest-assured:1.0
----
--

For more information about bootstrapping a Serverless Workflow project, see
xref:getting-started/create-your-first-workflow-service.adoc[Create Your First Serverless Workflow Service guide].

[[creating-workflow]]
== Creating a workflow

After bootstrapping a project, you need to create a workflow. In the following procedure, a workflow named
`Hello workflow` is created.

.Procedure
. Create a file named `hello.sw.json` in the `src/main/resources` directory with the following content:
+
--
.Example content for `hello.sw.json` file
[source,json]
----
{
  "id": "hello",
  "version": "1.0",
  "name": "Hello workflow",
  "description": "JSON based hello workflow",
  "start": "Hello",
  "states": [
    {
      "name": "Hello",
      "type": "inject",
      "data": {
        "hello": "Hello, "
      },
      "stateDataFilter": {
        "output": "${ {greeting: .hello+.name} }"
      },
      "end": true
    }
  ]
}
----
--

> For the workflow defined above, Kogito will create an endpoint at `/hello`, which is the `id` of the workflow.

Reviewer: ```suggestion
For the workflow defined above, the `/hello` will be created, which is the `id` of the workflow.
```

Reviewer 2: @spolti, for reading purposes, it's better to avoid passive voice, IMO.

What about:

```suggestion
The engine creates an endpoint at the `/hello` path. The path naming is the same as the workflow `id` in the definition file.
```

[[testing-application]]
== Testing your Serverless Workflow application

> After creating your workflow, you're ready to test your application. In the following procedure is created a
> test to check if the `hello` endpoint returns a proper greeting for a given `name`.

Reviewer 2: Same thing regarding passive voice:

```suggestion
After creating the workflow, you're ready to test the application. Follow the procedure below to create a test to check if the `hello` endpoint returns the proper greeting for a given `name`.
```

[INFO]
[INFO]
[INFO] --- maven-jar-plugin:2.4:jar (default-jar) @ serverless-workflow-testing-with-rest-assured ---
[INFO] Building jar: /home/helber/dev/kiegroup/kogito-examples/kogito-quarkus-examples/serverless-workflow-testing-with-rest-assured/target/serverless-workflow-testing-with-rest-assured-1.0.jar
[INFO]
[INFO] --- quarkus-maven-plugin:2.9.2.Final:build (default) @ serverless-workflow-testing-with-rest-assured ---
[INFO] Performed addonsConfig discovery, found: AddonsConfig{usePersistence=false, useTracing=false, useMonitoring=false, usePrometheusMonitoring=false, useCloudEvents=true, useExplainability=false, useProcessSVG=false, useEventDrivenDecisions=false, useEventDrivenRules=false}
[INFO] Generator discovery performed, found [processes]
[INFO] [io.quarkus.deployment.QuarkusAugmentor] Quarkus augmentation completed in 2669ms
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  15.558 s
[INFO] Finished at: 2022-06-03T16:36:19-03:00
[INFO] ------------------------------------------------------------------------
----
--

> == Additional resources

Reviewer 2: Hmm we have a few [IT related guides](https://github.com/kiegroup/kogito-docs/tree/main/serverlessworkflow/modules/ROOT/pages/testing-and-troubleshooting).

The idea is to add the "What's next" section here and link a few like "Mocking OpenAPI Services with Wiremock" and the Postgres you mentioned!

and https://marketplace.visualstudio.com/items?itemName=redhat.vscode-extension-serverless-workflow-editor[Red Hat Serverless Workflow Editor] is installed to edit your workflows.

For more information about the tooling and the required dependencies, see xref:getting-started/getting-familiar-with-our-tooling.adoc[tooling guide].

This guide describes how to test the Serverless Workflow application that exposes the `hello` endpoint.
The endpoint expects a `name` as a parameter and returns a `greeting`.

Example: For a request with the given body:

[source,json]
----
{
  "workflowdata": {
    "name": "John Doe"
  }
}
----

The endpoint should return a response with a body similar to:

[source,json]
----
{
  "id": "00ce5c1d-4936-4e18-998c-f5c5d5d784ad",
  "workflowdata": {
    "greeting": "Hello, John Doe"
  }
}
----

[[boostrapping-the-project]]
== Bootstrapping a project

To create your Serverless Workflow service you need to bootstrap a project. The example below is just to illustrate the testing example. Please see the xref:getting-started/create-your-first-workflow-service.adoc[Create Your First Serverless Workflow Service guide] for more information.

> .Procedure

Reviewer 3: Shouldn't we just refer to the getting started guide for people to get a working example first?  So this guide would start with the assumption that you have a working SW project, either because you executed the getting started guide, you opened up an example (for example the one you reference here at the top), or they have their own working example.  And the guide just starts by saying what you need to do to add a test.  The example test you show will match the example on kogito-examples, but users should be able to do something similar for the other use cases.  Wdyt?

Author: Thanks for the review, @krisv.
I pushed the requested change.
